File: impl-large/php/src/Domain.php
<?php
// Task Service, large tier — domain types, constants and pure rules.

declare(strict_types=1);

const MAX_TITLE_LENGTH = 80;
const MAX_NAME_LENGTH = 60;
const MAX_COMMENT_LENGTH = 200;
const MAX_BULK_ITEMS = 20;
const MIN_PRIORITY = 1;
const MAX_PRIORITY = 5;
const DEFAULT_LIMIT = 20;
const MAX_LIMIT = 100;
const DEFAULT_QUOTA = 10000;
const PROBE_QUOTA = 5;

const ROLES = ['admin', 'user'];
const STATUSES = ['todo', 'in_progress', 'done', 'archived'];
const STATUS_BONUS = ['todo' => 0, 'in_progress' => 3, 'done' => 5, 'archived' => 0];
const TRANSITIONS = ['todo->in_progress', 'todo->archived', 'in_progress->todo',
    'in_progress->done', 'done->archived'];
const PROJECT_SORTS = ['id', 'name', 'taskCount'];
const TASK_SORTS = ['id', 'title', 'priority', 'score', 'status'];
const USER_SORTS = ['id', 'username', 'role'];
const COMMENT_SORTS = ['id', 'authorId'];
const SEQ_SORTS = ['seq'];
const GROUP_BYS = ['assignee', 'status', 'project'];

final class User
{
    public function __construct(
        public int $id,
        public string $username,
        public string $password,
        public string $role,
        public int $quota,
        public int $version = 1,
        public bool $deleted = false,
    ) {
    }
}

final class Session
{
    public function __construct(
        public string $token,
        public int $userId,
        public int $used = 0,
    ) {
    }
}

final class Project
{
    public function __construct(
        public int $id,
        public string $name,
        public int $ownerId,
        public int $version = 1,
        public bool $deleted = false,
    ) {
    }
}

final class Task
{
    public function __construct(
        public int $id,
        public int $projectId,
        public string $title,
        public int $priority,
        public string $status,
        public ?int $assigneeId,
        public string $internalNote = '',
        public int $version = 1,
        public bool $deleted = false,
    ) {
    }
}

final class Comment
{
    public function __construct(
        public int $id,
        public int $taskId,
        public int $authorId,
        public string $body,
    ) {
    }
}

final class AuditEntry
{
    public function __construct(
        public int $seq,
        public int $actorId,
        public string $action,
        public string $resource,
        public int $resourceId,
    ) {
    }
}

final class OutboxEvent
{
    public function __construct(
        public int $seq,
        public string $name,
        public int $resourceId,
        public bool $delivered = false,
    ) {
    }
}

/** Every failure path throws this. The api layer turns it into the envelope. */
final class AppError extends Exception
{
    /** @param list<array{field: string, message: string}> $details */
    public function __construct(
        public int $status,
        public $code,
        public $message,
        public array $details = [],
    ) {
    }
}

function badRequest(): AppError
{
    return new AppError(400, 'bad_request', 'the request is malformed');
}

function unauthorized(): AppError
{
    return new AppError(401, 'unauthorized', 'authentication is required');
}

function invalidCredentials(): AppError
{
    return new AppError(401, 'invalid_credentials', 'the username or password is wrong');
}

function forbidden(): AppError
{
    return new AppError(403, 'forbidden', 'you may not access this resource');
}

function notFound(): AppError
{
    return new AppError(404, 'not_found', 'the resource does not exist');
}

function conflict(): AppError
{
    return new AppError(409, 'conflict', 'the resource already exists');
}

function invalidTransition(): AppError
{
    return new AppError(409, 'invalid_transition', 'the status change is not allowed');
}

function preconditionFailed(): AppError
{
    return new AppError(412, 'precondition_failed', 'the resource has changed');
}

function preconditionRequired(): AppError
{
    return new AppError(428, 'precondition_required', 'the If-Match header is required');
}

function quotaExceeded(): AppError
{
    return new AppError(429, 'quota_exceeded', 'the request quota is exhausted');
}

function invalid(array $details): AppError
{
    usort($details, fn (array $a, array $b) => [$a['field'], $a['message']]
        <=> [$b['field'], $b['message']]);
    return new AppError(422, 'validation_failed', 'the request body is not valid', $details);
}

/** @return array{field: string, message: string} */
function fail(string $field, string $message): array
{
    return ['field' => $field, 'message' => $message];
}

function computeScore(int $priority, string $status): int
{
    $baseScore = $priority * 10;
    return $baseScore + STATUS_BONUS[$status];
}

function checkString(string $value, string $fieldName, int $maxLength, array &$errors): void
{
    if ($value === '') {
        $errors[] = fail($fieldName, "$fieldName is required");
    } elseif (mb_strlen($value) > $maxLength) {
        $errors[] = fail($fieldName, "$fieldName is too long");
    }
}

function checkPriority(?int $value, array &$errors): void
{
    if ($value === null || $value < MIN_PRIORITY || $value > MAX_PRIORITY) {
        $errors[] = fail('priority', 'priority is out of range');
    }
}

function checkStatus(mixed $value, array &$errors): void
{
    if (!is_string($value) || !in_array($value, STATUSES, true)) {
        $errors[] = fail('status', 'status is not valid');
    }
}

function checkRole(mixed $value, array &$errors): void
{
    if (!is_string($value) || !in_array($value, ROLES, true)) {
        $errors[] = fail('role', 'role is not valid');
    }
}

function checkQuota(mixed $value, array &$errors): void
{
    if (!is_int($value) || $value < 0) {
        $errors[] = fail('quota', 'quota is out of range');
    }
}

File: impl-large/go/domain.go
// Task Service, large tier — domain types, constants and pure rules.
package main

import (
	"cmp"
	"encoding/json"
	"slices"
	"strings"
	"unicode/utf8"
)

const (
	maxTitleLength   = 80
	maxNameLength    = 60
	maxCommentLength = 200
	maxBulkItems     = 20
	minPriority      = 1
	maxPriority      = 5
	defaultLimit     = 20
	maxLimit         = 100
	defaultQuota     = 10000
	probeQuota       = 5
	port             = "127.0.0.1:8080"
)

var (
	roles       = []string{"admin", "user"}
	statuses    = []string{"todo", "in_progress", "done", "archived"}
	statusBonus = map[string]int{"todo": 0, "in_progress": 3, "done": 5, "archived": 0}
	transitions = [][2]string{
		{"todo", "in_progress"}, {"todo", "archived"}, {"in_progress", "todo"},
		{"in_progress", "done"}, {"done", "archived"},
	}
	projectSorts = []string{"id", "name", "taskCount"}
	taskSorts    = []string{"id", "title", "priority", "score", "status"}
	userSorts    = []string{"id", "username", "role"}
	commentSorts = []string{"id", "authorId"}
	seqSorts     = []string{"seq"}
	groupBys     = []string{"assignee", "status", "project"}
)

type User struct {
	ID       int
	Username string
	Password string
	Role     string
	Quota    int
	Version  int
	Deleted  bool
}

type Session struct {
	Token  string
	UserID int
	Used   int
}

type Project struct {
	ID      int
	Name    string
	OwnerID int
	Version int
	Deleted bool
}

type Task struct {
	ID           int
	ProjectID    int
	Title        string
	Priority     int
	Status       string
	AssigneeID   *int
	InternalNote string
	Version      int
	Deleted      bool
}

type Comment struct {
	ID       int
	TaskID   int
	AuthorID int
	Body     string
}

type AuditEntry struct {
	Seq        int
	ActorID    int
	Action     string
	Resource   string
	ResourceID int
}

type OutboxEvent struct {
	Seq        int
	Name       string
	ResourceID int
	Delivered  bool
}

// row is one serialized resource. A map keeps a conditional key honest.
type row = map[string]any

// ordered is a JSON object that keeps the order its keys were added in.
type ordered []member

type member struct {
	Key   string
	Value any
}

func (o ordered) MarshalJSON() ([]byte, error) {
	out := []byte("{")
	for index, entry := range o {
		key, err := json.Marshal(entry.Key)
		if err != nil {
			return nil, err
		}
		value, err := json.Marshal(entry.Value)
		if err != nil {
			return nil, err
		}
		if index > 0 {
			out = append(out, ',')
		}
		out = append(append(append(out, key...), ':'), value...)
	}
	return append(out, '}'), nil
}

type detail struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

// AppError is every failure path. The api layer turns it into the envelope.
type AppError struct {
	Status  int
	Code    string
	Message string
	Details []detail
}

func (e *AppError) Error() string { return e.Message }

func badRequest() *AppError {
	return &AppError{Status: 400, Code: "bad_request", Message: "the request is malformed"}
}

func unauthorized() *AppError {
	return &AppError{Status: 401, Code: "unauthorized", Message: "authentication is required"}
}

func invalidCredentials() *AppError {
	return &AppError{Status: 401, Code: "invalid_credentials",
		Message: "the username or password is wrong"}
}

func forbidden() *AppError {
	return &AppError{Status: 403, Code: "forbidden", Message: "you may not access this resource"}
}

func notFound() *AppError {
	return &AppError{Status: 404, Code: "not_found", Message: "the resource does not exist"}
}

func conflict() *AppError {
	return &AppError{Status: 409, Code: "conflict", Message: "the resource already exists"}
}

func invalidTransition() *AppError {
	return &AppError{Status: 409, Code: "invalid_transition",
		Message: "the status change is not allowed"}
}

func preconditionFailed() *AppError {
	return &AppError{Status: 412, Code: "precondition_failed",
		Message: "the resource has changed"}
}

func preconditionRequired() *AppError {
	return &AppError{Status: 428, Code: "precondition_required",
		Message: "the If-Match header is required"}
}

func quotaExceeded() *AppError {
	return &AppError{Status: 429, Code: "quota_exceeded",
		Message: "the request quota is exhausted"}
}

func invalid(details []detail) *AppError {
	slices.SortFunc(details, func(a, b detail) int {
		return cmp.Or(strings.Compare(a.Field, b.Field), strings.Compare(a.Message, b.Message))
	})
	return &AppError{Status: 422, Code: "validation_failed",
		Message: "the request body is not valid", Details: details}
}

func fail(field, message string) detail {
	return detail{Field: field, Message: message}
}

// pointer boxes a value so an absent field and a null field stay distinct.
func pointer[T any](value T) *T { return &value }

// value unwraps an optional integer, using zero so a lookup fails cleanly.
func value(number *int) int {
	if number == nil {
		return 0
	}
	return *number
}

func computeScore(priority int, status string) int {
	baseScore := priority * 10
	return baseScore + statusBonus[status]
}

func checkString(text, fieldName string, maxLength int, errors *[]detail) {
	if text == "" {
		*errors = append(*errors, fail(fieldName, fieldName+" is required"))
	} else if utf8.RuneCountInString(text) > maxLength {
		*errors = append(*errors, fail(fieldName, fieldName+" is too long"))
	}
}

func checkPriority(priority *int, errors *[]detail) {
	if priority == nil || *priority < minPriority || *priority > maxPriority {
		*errors = append(*errors, fail("priority", "priority is out of range"))
	}
}

func checkStatus(status *string, errors *[]detail) {
	if status == nil || !slices.Contains(statuses, *status) {
		*errors = append(*errors, fail("status", "status is not valid"))
	}
}

func checkRole(role *string, errors *[]detail) {
	if role == nil || !slices.Contains(roles, *role) {
		*errors = append(*errors, fail("role", "role is not valid"))
	}
}

func checkQuota(quota *int, errors *[]detail) {
	if quota == nil || *quota < 0 {
		*errors = append(*errors, fail("quota", "quota is out of range"))
	}
}
